### ISSUE TITLE
Running on multiple containers?

### ISSUE THREAD

**Issue by lostphilosopher** [moove-it/sidekiq-scheduler#223]
I'm running Sidekiq + Sidekiq Scheduler on OSE (v3). I've got a single Sidekiq pod with multiple containers. Each container has its own set of ENV vars. I only have one Sidekiq job. I want every container to run that job (starting at midnight via `cron:` _UPDATE: This won't work, I've switched to `every:`._) and then use their ENV vars to operate on a different subset of a large data set. However, when midnight comes around every container runs the same job - on the same set of data - simultaneously. (_UPDATE: This was false, only one container ran the sidekiq job, the logs were just no properly separated._)

I don't understand why that's true, or how to fix it. Can anyone help?

I think what I'm looking for is for each container to run its own Sidekiq job with its own job ID. They would all be running the same code (with different ENV vars) and using the same Redis, but each job should be independent of the others. 

SO Cross Reference: https://stackoverflow.com/questions/46713223/sidekiq-scheduled-jobs-in-multiple-containers

**Comment by lostphilosopher** [comment]
gem 'sidekiq', '4.1.4'
gem 'sidekiq-scheduler', '~> 2.0'

**Comment by lostphilosopher** [comment]
Might be this: https://github.com/moove-it/sidekiq-scheduler#notes-about-running-on-multiple-hosts?

**Comment by snmgian** [comment]
Where is Redis running? Does each container run its own Redis process, and each sidekiq is connected to that Redis?

**Comment by lostphilosopher** [comment]
There is a single instance of Redis that all the containers talk to.

**Comment by lostphilosopher** [comment]
Ok, switched to `every: '24h'`, the job was kicked off simultaneously on 4/8 containers and ran successfully there. The other 4 containers logged 0 Sidekiq activity.

**Comment by snmgian** [comment]
It's ok for `cron` jobs to run only *once*, no matter how many containers you have. The rational here is that if you have multiple containers for load-balancing, you'll want the job to be run only once.

Also, as sidekiq instances are sharing the same Redis, you're not guaranteed that each container will run a job (but see option 2)

Options here are:
 - Have a redis instance per container.
 - Have a separate queue for each container. Take a llok at: https://github.com/mperham/sidekiq/wiki/FAQ#how-do-i-ensure-a-job-processes-on-a-given-machine

**Comment by lostphilosopher** [comment]
@snmgian ahhh, ok, things are coming together a bit... I've modified my sidekiq.yml:
```
---
:verbose: false
:concurrency: 25
:logfile: ./log/sidekiq-<%= ENV["MOD_REMAINDER"] || 0 %>.log
:queues:
  - sidekiq-<%= ENV["MOD_REMAINDER"] || 0 %>
```
And my schedule is till using `every`:
```
CacheWarmer:
#  every day at midnight UTC (7pm central)
  every: ["24h", first_at: "2017-10-16T00:00:00.000+00:00"]
```

**Comment by snmgian** [comment]
You should explicitly set to which queue sidekiq-scheduler should send the jobs:

```yaml
CacheWarmer:
#  every day at midnight UTC (7pm central)
  every: ["24h", first_at: "2017-10-16T00:00:00.000+00:00"]
  queue: sidekiq-<%= ENV["MOD_REMAINDER"] || 0 %>
```

Maybe you're already doing this ...

**Comment by lostphilosopher** [comment]
@snmgian that ^ doesn't seem to work. Here's the latest code:

sidekiq.yml
```
---
:verbose: false
:concurrency: 25
:logfile: ./log/sidekiq-<%= ENV["MOD_REMAINDER"] || 0 %>.log
:queues:
  - sidekiq-<%= ENV["MOD_REMAINDER"] || 0 %>
```

worker_schedules.yml
```
CacheWarmer:
#  every day at midnight UTC (7pm central)
  every: ["24h", first_at: "<%= Time.now.midnight %>"]
  queue: sidekiq-<%= ENV["MOD_REMAINDER"] || 0 %>
```

scheduler.rb (scheduler initializer)
```
require 'sidekiq/scheduler'

Sidekiq.configure_server do |config|
  config.on(:startup) do
    Sidekiq.schedule = YAML.load(ERB.new(File.read("#{Rails.root}/config/worker_schedules.yml")).result)
    Sidekiq::Scheduler.reload_schedule!
  end
end
```

When I'm ssh'd into any of my sidekiq containers `ENV["MOD_REMAINDER"]` reports the right value, and if I run the code from the initializer and look at `Sidekiq.get_schedule` it will look right, but I if ssh into a *different* container `Sidekiq.get_schedule` reports the values set by the previous container - all the containers instances of SidekiqScheduler write to the same queue.

**Comment by lostphilosopher** [comment]
My next move will be to have sidekiq in each container initialize with a different Redis DB on the same Redis instance (then queue shouldn't matter since each DB will have its own default queue).

**Comment by snmgian** [comment]
By default, Redis has 16 DBs, keep that in mind if you have more than 16 containers. Of course, you can configure the number of DBs.

I see now why the suffixed queue names approach doesn't work ... job config is stored in redis, so the last container that attempts to store the config is the one who wins.

Separated Redis instances / separate DBs for each container would be simpler approach.

Are you running one Sidekiq Web per container?

**Comment by lostphilosopher** [comment]
>" job config is stored in redis, so the last container that attempts to store the config is the one who wins."

Yep, that was my conclusion as well. Switching to a separate Redis DB `/0`, `/1`, `/2`, etc. Looks like it worked from a parallelization perspective. (`/0` is a bit out of whack because I think it's still picking up the old jobs). I have 8 containers so for now I'm ok on Redis DBs.

> "Are you running one Sidekiq Web per container?"

I'm not sure. The web console doesn't seem to be working. I can see the jobs running currently while ssh'd into the container, but /sidekiq reports 0 busy processes. I hadn't really thought about how run the built in monitoring in this configuration... Do you have a recommendation?

**Comment by snmgian** [comment]
Regarding 0 busy processes report, the problem could be to not configuring Redis DB. Take a look at: https://github.com/mperham/sidekiq/wiki/Monitoring#standalone

Example:
```ruby
require 'sidekiq'

Sidekiq.configure_client do |config|
  :url => "#{ENV["REDIS_URL"]}#{ENV["MOD_REMAINDER"] || 0}"
end

require 'sidekiq/web'
require 'sidekiq-scheduler/web'
run Sidekiq::Web
```

**Comment by snmgian** [comment]
> (/0 is a bit out of whack because I think it's still picking up the old jobs)

Have you verified that by looking at the logs of that container?

**Comment by lostphilosopher** [comment]
@snmgian I got caught up in some other projects, but I wanted to update here. The latest configuration (unique Redis DB per container) is working great. Everything seems to be working as expected. I've still got work to do on the monitoring, but I've added extensive logging that is meeting my needs for now. Thank you for your help! You're welcome to consider this issue closed, though I may make an update later on regarding the monitoring part.
